src/ripple/protocol/STObject.h

```diff
@@ -734,6 +761,22 @@ STObject::ValueProxy<T>::operator= (U&& u)
     return *this;
 }
 
+template <class T>
+template <class Unit, class U >
+std::enable_if_t<
+    std::is_assignable_v<T, U> &&
+    (
+    std::is_same_v<Unit, units::drop_unit> ||
```

Comment:
do we actually need to restrict `Unit` to these three types? I'll admit I don't fully understand this code, but I was hoping the assign from argument type would be sufficient.

Reply:
I consider this an important safety mechanism because we're stripping out unit information to make the assignment. Mechanically, the main intention is to keep values with "intermediate" or "composed" units from incorrectly being assigned into objects and losing that unit information. Consider the following pseudocode:
```
STObject o = [whatever];
auto x = 150 * units::drops;
auto y = 10 * units::feeunits;

o[sfField1] = x; // This is good. x is in drops, so it makes sense to allow this assignment. sfField1 is 150.
o[sfField2] = y; // This is also good. y is fee units, so it makes sense to allow this assignment. sfField2 is 10.
o[sfField3] = x * y; // This is problematic. The unit of x*y is basically "fee_unit drops". sfField3 is 1500, but 1500 what?
o[sfField4] = x / y; // Wat. The unit of x/y is basically "drops/fee_units". sfField4 has a value of 15, but again 15 what?
```
So the intention of this restriction is to prevent assignments like sfField3 & 4 from happening by accident. If one ever needs to be done intentionally, then we can add that case later. For now, I'd rather it be restrictive.